### ISSUE TITLE
Collection count is inconsistent

### ISSUE THREAD

**Issue by gabata** [arangodb/arangodb#5695]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [ ] 2.8
- [ ] 3.0
- [ ] 3.1
- [ ] 3.2
- [x] 3.3 (`ArangoDB 3.3.10 [linux] 64bit, using jemalloc, build tags/v3.3.10-0-g902a980389, VPack 0.1.30, RocksDB 5.6.0, ICU 58.1, V8 5.7.492.77, OpenSSL 1.1.0f 25 May 2017`)
- [ ] self-compiled devel branch

Mode:
- [ ] Cluster
- [x] Single-Server

Storage-Engine:
- [ ] mmfiles
- [x] rocksdb

On this operating system:
- [ ] DCOS on
  - [ ] AWS
  - [ ] Azure
  - [ ] own infrastructure
- [x] Linux 
  - [ ] Debian .deb
  - [ ] Ubuntu .deb
  - [ ] SUSE   .rpm
  - [ ] RedHat .rpm
  - [ ] Fedora .rpm
  - [ ] Gentoo
  - [x] docker - official docker library (detected operating system: Linux version 4.14.33-51.37.amzn1.x86_64 (`mockbuild@gobi-build-60006) (gcc version 7.2.1 20170915 (Red Hat 7.2.1-2) (GCC)) #1 SMP Thu May 3 20:07:43 UTC 2018`)
  - [ ] other:
- [ ] Windows, version: 
- [ ] MacOS, version:

### this is an AQL-related issue:
- [x] I'm using graph features

I'm issuing AQL via:
- [x] web interface with this browser: Firefox    running on this OS: Windows 10 x64
- [ ] arangosh
- [ ] this Driver:

I wanted to copy a document collection into an other empty collection via the following AQL

```
FOR doc IN `collection1`
INSERT doc IN `collection2`
OPTIONS { ignoreErrors: true }
```
I'm using user-defined keys, and my documents look like the following:

```javascript
{
  _id: "collection1/DUMMY_TYPE_NAME.id:123456789.brand:xyz",
  _key: "DUMMY_TYPE_NAME.id:123456789.brand:xyz",
  "dataSource": "test data source",
  "ids": {
    "id": "123456789",
    "brand": "xyz"
  },
  "modificationDate": "2018-06-25T23:38:56.754Z",
  "productType": "DUMMY_TYPE_NAME",
  "score1": 10,
  "score2": 0,
  "score3": 5
}
```

After the AQL has been executed I have experienced that the new collection contains ~5x more documents than the original collection:

```javascript
$ var sourceCollection = db._collection("collection1");
$ var targetCollection = db._collection("collection2");
$ print(sourceCollection.count(), targetCollection.count());
3217293 17102695
```
The Arango web UI shows the same numbers.

Could you please help what can be the problem here?

**Comment by jsteemann** [comment]
I just tried the following script:
```js
let doc = {
  _id: "collection1/DUMMY_TYPE_NAME.id:123456789.brand:xyz",
  _key: "DUMMY_TYPE_NAME.id:123456789.brand:xyz",
  "dataSource": "test data source",
  "ids": {
    "id": "123456789",
    "brand": "xyz"
  },
  "modificationDate": "2018-06-25T23:38:56.754Z",
  "productType": "DUMMY_TYPE_NAME",
  "score1": 10,
  "score2": 0,
  "score3": 5
};
db._create("collection1");
db._create("collection2");

for (let i = 0; i < 3 * 1000 * 1000; ++i) {
  doc._key = "DUMMY_TYPE_NAME.id:" + i + ".brand:xyz";
  db.collection1.insert(doc);
}

db._query("FOR doc IN `collection1` INSERT doc IN `collection2` OPTIONS { ignoreErrors: true }");

let sourceCollection = db._collection("collection1");
let targetCollection = db._collection("collection2");

print(sourceCollection.count(), targetCollection.count());
```
and it printed `10000 10000`.

A few questions:
* are there any other operations happening on the target collection that could change the document count
* why are you explicitly setting `OPTIONS { ignoreErrors: true }` for the query? Does it mean something goes wrong during the insert or is expected to go wrong?

**Comment by gabata** [comment]
@jsteemann thanks for checking.

> are there any other operations happening on the target collection that could change the document count

There are no other operations which could change the target collection count.

> why are you explicitly setting `OPTIONS { ignoreErrors: true }` for the query? Does it mean something goes wrong during the insert or is expected to go wrong?

Otherwise I get the following error message `Query: AQL: unique constraint violated (while executing)` which must be impossible in my opinion. It seems that the database is somehow in inconsistent state.

**Comment by jsteemann** [comment]
Are there any indexes present on the target collection, i.e. what does `db.collection2.getIndexes()` show?

**Comment by gabata** [comment]
There are no additional indices,  only the build-in index for `_key`:
```
[
  {
    "fields" : [
      "_key"
    ],
    "id" : "collection2/0",
    "selectivityEstimate" : 1,
    "sparse" : false,
    "type" : "primary",
    "unique" : true
  }
]
```

**Comment by jsteemann** [comment]
Does the problem occur reproducibly, i.e. if you drop collection2 and re-create it, does it work afterwards or fail again?

**Comment by jsteemann** [comment]
Can you run `db.collection1.properties()` and `db.collection2.properties()` and share the output here, plus the output of the AQL query `FOR doc IN collection1 COLLECT k = doc._key WITH COUNT INTO c FILTER c > 1 RETURN [k, c]`? The query should produce an empty result, as there is the unique index on _key.

**Comment by gabata** [comment]
```
`db.collection1.properties()` and `db.collection2.properties()` have the same result:
{                                
  "isSystem" : false,            
  "waitForSync" : false,         
  "keyOptions" : {               
    "type" : "traditional",      
    "allowUserKeys" : true,      
    "lastValue" : 0              
  },                             
  "cacheEnabled" : false         
}                                
```

The result of `FOR doc IN collection1 COLLECT k = doc._key WITH COUNT INTO c FILTER c > 1 RETURN [k, c]` is disappointing (have masked the key values due to our NDA):

```
[
  [
    "TYPE_A.id:10519.type:15",
    240
  ],
  [
    "TYPE_A.id:179471.type:3",
    302
  ],
  [
    "TYPE_A.id:180023.type:12",
    280
  ],
  [
    "TYPE_A.id:601874.type:12",
    326
  ],
  [
    "TYPE_A.id:602503.type:37",
    129
  ],
  [
    "TYPE_H.id:10019",
    2357
  ],
  [
    "TYPE_H.id:10020",
    1789
  ],
  [
    "TYPE_H.id:10043",
    1568
  ],
  ...
```

**Comment by jsteemann** [comment]
Ok, then the AQL INSERT statement behaves correctly when throwing "unique constraint violation" errors.
The problem seems to be with the source collection having multiple same _key values.
Do you remember how the source collection was created (e.g. via restore, via import, via db._create...) and how the documents in it were generated?

**Comment by jsteemann** [comment]
Ok, thanks. This helps. Will need time for investigation though.

**Comment by graetzer** [comment]
Hey @gabata,

I tried to reproduce this with the import API on version 3.3 with multiple works threads and various settings for `onDuplicate` but was not successfull :-(.
Is there any more information here that might be useful for us? Perhaps you can give us some more details about your program that would help us to reproduce this collection corruption  ?

**Comment by graetzer** [comment]
ok we figured out that this is related to the intermediate commit size https://docs.arangodb.com/devel/Manual/Administration/Configuration/RocksDB.html#non-pass-through-options

**Comment by gabata** [comment]
Hi @graetzer thanks for analyzing.

We have the following RocksDB config:

```bash
--rocksdb.max-transaction-size #{2 * 1024 ** 3}      # transaction size limit (in bytes) (2 GB)
--rocksdb.intermediate-commit-size #{64 * 1024 ** 2} # threshold of auto-commit and start a new transaction (bytes) (64 MB)
--rocksdb.intermediate-commit-count 1000             # threshold of auto-commit and start a new transaction (ops count)
--rocksdb.write-buffer-size #{64 * 1024 ** 2}        # amount of data to build up in memory before converting to file (64 MB)
--rocksdb.block-cache-size #{10 * 1024 ** 3}         # block cache size in bytes (10 GB)
```

Do you have any suggestions how we can avoid the situation?

**Comment by graetzer** [comment]
Sorry to get back to you so late: we are going to test out a solution for this in the devel branch soon and then backport it. Currently the best option is to set `--rocksdb.intermediate-commit-count` and `--rocksdb.intermediate-commit-size` to high values to avoid triggering an intermediate commit

**Comment by maxkernbach** [comment]
Hi @gabata,

sorry for the late reply. Intermediate commits have been addressed with version 3.3.14 (PR https://github.com/arangodb/arangodb/pull/6039):

> Intermediate commits in the RocksDB engine are now only enabled in standalone AQL queries 
  (not within a JS transaction), standalone truncate as well as for the "import" API

Could you please check if this issue is fixed in version 3.3.14+ ? The latest 3.3 release is 3.3.16.

**Comment by gabata** [comment]
Hi @maxkernbach, I could not reproduce the issue in 3.3.14. Thanks for the fix, I'm closing this issue.
